src/Microsoft.OData.Core/UrlValidation/ValidationEngine/ODataUrlValidator.cs

```diff
@@ -413,14 +413,21 @@ private void ValidateImpliedProperties(IEdmType segmentType, SelectExpandClause
         // Validate all structured properties of a type, recursing through nested complex typed properties
         private void ValidateProperties(IEdmType edmType, ODataUrlValidationContext context)
         {
-            IEdmStructuredType structuredType = edmType as IEdmStructuredType;
-            if (structuredType != null)
+            if (!context.ValidatedTypes.Contains(edmType))

```

Comment:
>(!context.ValidatedTypes.Contains(edmType) [](start = 15, length = 42)

ISet.Add returns false if the item was already in the list, so you can replace 416 with:

if (context.ValidatedTypes.Add(edmType))

and remove line 418, avoiding the extra look-up.

Reply:
#Resolved